### PR TITLE
Data archival task implementation to prune CDM transnational tables.

### PR DESCRIPTION
**Pull request by dunithd**
In a production deployment of IoT server, activities and relevant data keep accumulating over time. So tables such as DM_OPERATION, DM_DEVICE_OPERATION_RESPONSE, and DM_ENROLMENT_OP_MAPPING will grow in size eventually. Due to this reason, reading activities will get slower.
To address this, a scheduled task needs to implemented to run once a day and move all records that are older than 30 days to another set of archival tables. This helps the system to retain activities only for the last month and will make the getActivities() endpoint faster to some extent.

There will be two scheduled tasks, both run exactly once a day.
Task #1 performs following on DM_OPERATION, DM_DEVICE_OPERATION_RESPONSE and DM_ENROLMENT_OP_MAPPING tables.

Read all records that are older than 30 days.
Insert them into corresponding tables in a new database. Those tables will be suffixed with "_arch".
From source tables, delete all records that are older than 30 days.
Task #2 performs following on the tables with "_arch" prefix.

Read all records that are older than 30 days( this retention period can be customized) and delete them.
These operations will be implemented with atomicity in mind. During the data move, records will be broken into a set of batches and they will be inserted and deleted in a transactional manner, batch after batch. We'll have to do some load tests and determine a suitable value for the batch size.

Related issue:https://github.com/wso2/product-iots/issues/1394

### CODE REVIEW

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/dao/ArchivalDestinationDAOFactory.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
+package org.wso2.carbon.device.mgt.core.archival.dao;
```
Comment (sinthuja) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/dao/ArchivalDestinationDAOFactory.java]: Missing license header.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/ArchivalService.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package org.wso2.carbon.device.mgt.core.archival;
```
Comment (sinthuja) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/ArchivalService.java]: License header should be on the top, before the package name.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/ArchivalException.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package org.wso2.carbon.device.mgt.core.archival;
```
Comment (sinthuja) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/ArchivalException.java:19]: Missing license header.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/ArchivalService.java (hunk 2 above)
Comment (sinthuja) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/ArchivalService.java]: License header should be at the top before the package name declaration.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/dao/ArchivalDAOException.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/*
+ *   Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (sinthuja) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/dao/ArchivalDAOException.java]: License header should be 2017.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/dao/ArchivalDAOUtil.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
+/*
+ *   Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (sinthuja) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/dao/ArchivalDAOUtil.java]: License header should be 2017.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/ArchivalTaskManagerImpl.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            String taskClazz = DeviceConfigurationManager.getInstance().getDeviceManagementConfig()
+                    .getArchivalConfiguration().getArchivalTaskConfiguration().getTaskClazz();
+            String cronExpression = DeviceConfigurationManager.getInstance().getDeviceManagementConfig()
+                    .getArchivalConfiguration().getArchivalTaskConfiguration().getCronExpression();
+
+            TaskManager taskManager = taskService.getTaskManager(TASK_TYPE_ARCHIVAL);
+
+            TaskInfo.TriggerInfo triggerInfo = new TaskInfo.TriggerInfo();
+            triggerInfo.setCronExpression(cronExpression);
+//            triggerInfo.setIntervalMillis(60000);
```
Comment (sinthuja) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/ArchivalTaskManagerImpl.java]: Remove the commented unwanted code.

File: features/device-mgt/org.wso2.carbon.device.mgt.server.feature/src/main/resources/dbscripts/cdm/mysql.sql [hunk 7]
```diff
@@ -589,3 +589,10 @@ DM_DEVICE.ID = DM_DEVICE_DETAIL.DEVICE_ID
 ORDER BY TENANT_ID, DEVICE_ID;
 
 -- END OF DASHBOARD RELATED VIEWS --
+
+-- TEMP TABLE REQUIRED FOR DATA ARCHIVAL JOB
+CREATE TABLE IF NOT EXISTS DM_ARCHIVED_OPERATIONS (
```
Comment (sinthuja) [inline features/device-mgt/org.wso2.carbon.device.mgt.server.feature/src/main/resources/dbscripts/cdm/mysql.sql]: Why only this table is included in the cdm, and the rest of the archival tables are in the archival folder?

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/dao/ArchivalDestinationDAOFactory.java (hunk 1 above)
Comment (dunithd) [reply components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/dao/ArchivalDestinationDAOFactory.java]: Fixed

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/ArchivalException.java (hunk 3 above)
Comment (dunithd) [reply components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/ArchivalException.java:19]: Fixed

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/ArchivalService.java (hunk 2 above)
Comment (dunithd) [reply components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/archival/ArchivalService.java]: Fixed

File: features/device-mgt/org.wso2.carbon.device.mgt.server.feature/src/main/resources/dbscripts/cdm/mysql.sql (hunk 7 above)
Comment (dunithd) [reply features/device-mgt/org.wso2.carbon.device.mgt.server.feature/src/main/resources/dbscripts/cdm/mysql.sql]: @sinthuja This is the temp table that is used when moving records to archival tables.

**Comment by rasika** [comment]
Hi dunith,
We are in the process of merging this PR. Can you resolve the conflicts of the PR and restructure the PR description into newest PR structure.

**Review by charithag — APPROVED** [review]

**Pull request merged**